### PR TITLE
feature(rome_js_parser): better handle of unicode for string literals, templates and regex

### PR DESCRIPTION
**Pull request by xunilrj**
We are doing something dangerous, that was the cause of https://github.com/rome/tools/pull/2344.

When parsing the contents of string literals, regexes (https://english.stackexchange.com/questions/508834/what-is-the-plural-form-of-regex) and string templates; we normally do

```rust
while let Some(byte) = self.next_byte() {
    ...
}
```

This is dangerous because sometimes we check with an ```if```, if we are in a linebreak. Although the Unicode committee try to avoid ambiguity when creating Unicode chars, we fall into traps because we also use ```from_utf8_unchecked``` out of char boundaries.

An example is the ‍ Zero Width Joiner Unicode char (https://unicode-table.com/en/200D/) that caused the #2344 bug. Its bytes are ```E2 80 8D```, and  ```from_utf8_unchecked``` will return an ````\r```` when we call it on its  second byte.

See this here:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=66e9d16029decb91cdfcb7a41f2757bf

Look at the 15th byte;

To avoid this issue this PR correctly iterate the code using utf8 chars when they are possible. I tried to keep ```next_byte``` where it is safe because it is faster.

### CODE REVIEW

**Comment by xunilrj** [comment]
!bench_parser

**Comment by Boshen** [comment]
I'm not sure if we should just rewrite the lexer using https://doc.rust-lang.org/stable/std/str/struct.Chars.html, all these boundary checking can be tucked away, and it shouldn't affect performance. 

I'm bringing this up because I studied this https://github.com/mozilla-spidermonkey/jsparagus/blob/master/crates/parser/src/lexer.rs#L27 , its tokenization code is so much cleaner, especially for unicode handling.

**Comment by xunilrj** [comment]
> I'm not sure if we should just rewrite the lexer ...

Yes. I am also wondering that. I actually have a "poc" to create a discussion on how to abstract encoding for the lexer. In this case, we would also support UTF-16 and others...

**Comment by ematipico** [comment]
On that note, I think we should make this generic and not isolated to a single parser. There are many places where we will need to handle strings with different encoding: different parsers, formatter, analysers, documentation generation (hence comments), etc.

**Comment by MichaReiser** [comment]
>  I'm not sure if we should just rewrite the lexer ...

Rewriting the lexer isn't a priority right now. It's working well enough and there are many other technical challenges that need to be tackled first. 

> On that note, I think we should make this generic and not isolated to a single parser. There are many places where we will need to handle strings with different encoding: different parsers, formatter, analysers, documentation generation (hence comments), etc.

The Lexer is very performance sensitive which is why a generic solution might not be feasible. I would even argue that any non UTF8 encoding should be converted to UTF8 before passing it to the lexer because non UTF8 is an edge case and we shouldn't sacrifice performance to handle the few non UTF8 files. 

Also, the `rust-cssparser` is working directly on bytes which is working fine. [source](https://github.com/servo/rust-cssparser/blob/master/src/tokenizer.rs). We probably just need to be more careful when we use `advance`, `next_byte` vs `next_char`.

**Comment by xunilrj** [comment]
!bench_parser

File: crates/rome_js_parser/src/lexer/mod.rs [hunk 1]
```diff
@@ -462,41 +463,54 @@ impl<'src> Lexer<'src> {
...
+                b'\r' | b'\n' => {
+                    self.advance(1);
+                }
+                chr => {
+                    if chr >= 0x80 {
+                        let chr = self.current_char_unchecked();
+                        if is_linebreak(chr) {
+                            self.advance(chr.len_utf8());
+                        }
+                    }
```
Comment (MichaReiser) [inline crates/rome_js_parser/src/lexer/mod.rs:480]: Can we abstract away this check? Why is it needed here but not when matching on `b'r'` directly?

**Comment by xunilrj** [comment]
!bench_parser

File: crates/rome_js_parser/src/lexer/mod.rs [hunk 2]
```diff
@@ -430,7 +428,11 @@ impl<'src> Lexer<'src> {
                             self.next_byte();
                         }
                         _ => {
-                            self.next_byte();
+                            if chr < 0x80 {
+                                self.next_byte();
+                            } else {
+                                self.advance_char_unchecked();
+                            }
```
[the remark this answers is not in the document]
Comment (xunilrj) [reply crates/rome_js_parser/src/lexer/mod.rs]: Done. Using ```is_ascii``` is more clear.

> Would it be easier to just always call advance_char?

Yes. But this check is faster because 99% of the time we don't have UTF8 chars.

File: crates/rome_js_parser/src/lexer/mod.rs [hunk 3]
```diff
@@ -462,54 +463,82 @@ impl<'src> Lexer<'src> {
...
+    ///
+    /// ## Safety
+    /// Must be called at a valid UT8 char boundary
+    fn consume_whitespaces(&mut self) {
+        self.assert_at_char_boundary();
+
+        while let Some(chr) = self.current_byte() {
+            match lookup_byte(chr) {
+                Dispatch::WHS => {
+                    if let b'\r' | b'\n' = chr {
```
Comment (MichaReiser) [inline crates/rome_js_parser/src/lexer/mod.rs:500]: Shouldn't this also test for the unicode line break characters (`is_linebreak`?)

File: crates/rome_js_parser/src/lexer/mod.rs [hunk 4]
```diff
@@ -462,54 +463,82 @@ impl<'src> Lexer<'src> {
...
+                    let chr = self.current_char_unchecked();
 
-            let dispatcher = lookup_byte(current);
-            if dispatcher == self::bytes::Dispatch::WHS
-                || (UNICODE_WHITESPACE_STARTS.contains(&current) && UNICODE_SPACES.contains(&chr))
-            {
-                self.advance(chr.len_utf8());
-            } else {
-                break;
+                    if UNICODE_SPACES.contains(&chr) {
```
Comment (MichaReiser) [inline crates/rome_js_parser/src/lexer/mod.rs:509]: How does this ensure that the `chr` is a whitespace?

File: crates/rome_js_parser/src/lexer/mod.rs [hunk 5]
```diff
@@ -540,6 +569,12 @@ impl<'src> Lexer<'src> {
         }
     }
 
+    /// Asserts that the lexer is at a UTF8 char boundary
+    #[inline]
+    fn assert_at_char_boundary(&self) {
```
Comment (MichaReiser) [inline crates/rome_js_parser/src/lexer/mod.rs:574]: Would it be safe to also add this chack to `advance`?

File: crates/rome_js_parser/src/lexer/mod.rs [hunk 6]
```diff
@@ -850,45 +882,48 @@ impl<'src> Lexer<'src> {
...
+                chr if chr == quote => {
+                    self.advance(1);
+                    return valid;
+                }
+                chr => {
+                    if chr.is_ascii() {
+                        self.advance(1);
+                    } else {
+                        self.advance_char_unchecked();
+                    }
```
Comment (MichaReiser) [inline crates/rome_js_parser/src/lexer/mod.rs:919]: Nit: How about moving this into a helper? `advance_byte_or_char`

File: crates/rome_js_parser/src/lexer/mod.rs [hunk 7]
```diff
@@ -1455,19 +1493,25 @@ impl<'src> Lexer<'src> {
...
+                    if let b'\r' | b'\n' = chr {
                         return COMMENT;
+                    } else if chr.is_ascii() {
+                        self.advance(1);
+                    } else {
+                        let chr = self.current_char_unchecked();
+                        if is_linebreak(chr) {
+                            return COMMENT;
+                        } else {
+                            self.advance(chr.len_utf8());
```
Comment (MichaReiser) [inline crates/rome_js_parser/src/lexer/mod.rs:1507]: When to use `advance(chr.len_utf8())` vs `advance_next_char_unchecked`?

File: crates/rome_js_parser/src/lexer/mod.rs [hunk 8]
```diff
@@ -1933,39 +2001,45 @@ impl<'src> Lexer<'src> {
...
-                }
-                Some(b'$') if self.peek_byte() == Some(b'{') => {
-                    token = Some(JsSyntaxKind::TEMPLATE_CHUNK);
-                    break;
-                }
-                Some(_) => {
-                    let _ = self.next_byte();
+                b'$' => {
+                    if let Some('{') = self.peek_char() {
+                        if self.position == start {
```
Comment (MichaReiser) [inline crates/rome_js_parser/src/lexer/mod.rs:2026]: Isn't `peek_byte` enough here because `{` is ascii?

**Review by MichaReiser — COMMENTED** [review]
Thanks for adding the asserts and the overall improvement of the lexer code. 

What are the changes to coverage submodules?

File: crates/rome_js_parser/src/lexer/mod.rs (hunk 3 above)
Comment (xunilrj) [reply crates/rome_js_parser/src/lexer/mod.rs:500]: No, because of the ```lookup_byte```. If it were a Unicode it would return an ```Dispatch::UNI```.

File: crates/rome_js_parser/src/lexer/mod.rs (hunk 4 above)
Comment (xunilrj) [reply crates/rome_js_parser/src/lexer/mod.rs:509]: Because UNICODE_SPACES contains all whitespaces we support.

What the old code does is check the first byteagainst ```UNICODE_WHITESPACE_STARTS``` and after the whole char at ```UNICODE_SPACES```, if we had a match at the first case.

I can skip the first byte check because it makes sense as optimization only for non-Unicode chars. But this case I moved to the above match arm.

File: crates/rome_js_parser/src/lexer/mod.rs (hunk 5 above)
Comment (xunilrj) [reply crates/rome_js_parser/src/lexer/mod.rs:574]: Ideally yes. But we still have a small problem with identifiers and keywords that I want to deal with later.

File: crates/rome_js_parser/src/lexer/mod.rs (hunk 7 above)
Comment (xunilrj) [reply crates/rome_js_parser/src/lexer/mod.rs:1507]: ```advance(chr.len_utf8())```: you already have the ```char```.
```advance_next_char_unchecked```: you don't have the ```char```.

The second has the cost of reading the buffer.

An option to simplify all these cases could be:

```
fn advance(step: impl LexerAdvance) {
    self.pos += step.step();
}

trait LexerAdvance {
    fn step(&self, buffer: &[u8]) -> usize;
}

impl LexerAdvance for usize {
    fn step(&self, buffer: &[u8]) -> usize { self }
}

impl LexerAdvance for u8 {
    fn step(&self, buffer: &[u8]) -> usize {
        if self.is_ascii() {
            1
        } else {
            let chr = unsafe { std::str::from_utf8_unchecked(buffer) }.chars().next().unwrap();
            chr.len_utf8()
        }
    }
}

impl LexerAdvance for u8 {
    fn step(&self, buffer: &[u8]) -> usize {
        self.len_utf8()
    }
}
```

all we need to do now is call ```self.advance(1)``` or ```self.advance(chr)``` for both ```u8``` and ```char```.
And the compiler will get the best case, inline and optimize everything away.

We need to be careful with what the compiler will call on ```self.advance(1)```. Maybe an enum to disambiguate will be the better option here:

```
self.advance(Step::OneByte)
```

File: crates/rome_js_parser/src/lexer/mod.rs (hunk 8 above)
Comment (xunilrj) [reply crates/rome_js_parser/src/lexer/mod.rs:2026]: Yes.

File: crates/rome_js_parser/src/lexer/mod.rs (hunk 3 above)
Comment (MichaReiser) [reply crates/rome_js_parser/src/lexer/mod.rs:500]: I see, and the line breaks aren't part of the `UNICODE_SPACES`, thus, the method breaks for line breaks. Neat

File: crates/rome_js_parser/src/lexer/mod.rs (hunk 8 above)
Comment (xunilrj) [reply crates/rome_js_parser/src/lexer/mod.rs:2026]: Done.

File: crates/rome_js_parser/src/lexer/mod.rs (hunk 6 above)
Comment (xunilrj) [reply crates/rome_js_parser/src/lexer/mod.rs:919]: Done.

**Review by ematipico — APPROVED** [review]
